cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/MessagingNotificationService.java

```diff
@@ -0,0 +1,180 @@
+    LOG.trace("Publishing on notification feed [{}]: {}", feed, notification);
+    return publishingExecutor.submit(new Callable<N>() {
+      @Override
+      public N call() throws Exception {
+        try {
+          NotificationMessage message = new NotificationMessage(feed, GSON.toJsonTree(notification, notificationType));
+          messagingService.publish(StoreRequestBuilder.of(notificationTopic).addPayloads(GSON.toJson(message)).build());
+          return notification;
+        } catch (Exception e) {
+          throw new NotificationException(e);
+        }
+      }
+    });
+  }
+
+  @Override
+  public <N> Cancellable subscribe(NotificationFeedId feed, NotificationHandler<N> handler,
+                                   Executor executor)
+    throws NotificationFeedNotFoundException, NotificationFeedException {
+
+    Cancellable subscribeCancellable = super.subscribe(feed, handler, executor);
+
+    // If already has a thread fetching, just return the cancellable.
+    if (!needFetch.compareAndSet(false, true)) {
+      return subscribeCancellable;
+    }
+
+    // Start fetching
+    subscribeExecutor.execute(new Runnable() {
+
+      private final long startTime = System.currentTimeMillis();
+      private final RetryStrategy scheduleStrategy = RetryStrategies.exponentialDelay(100, 3000, TimeUnit.MILLISECONDS);
+      private byte[] messageId;
+      private int emptyFetchCount;
+
+      @Override
+      public void run() {
+        try {
+          MessageFetcher fetcher = messagingService.prepareFetch(notificationTopic);
+          if (messageId == null) {
+            fetcher.setStartTime(startTime);
+          } else {
+            fetcher.setStartMessage(messageId, false);
+          }
+
+          emptyFetchCount++;
+          try (CloseableIterator<RawMessage> iterator = fetcher.fetch()) {
+            while (iterator.hasNext()) {
+              emptyFetchCount = 0;
+              RawMessage rawMessage = iterator.next();
+              NotificationMessage message = GSON.fromJson(new String(rawMessage.getPayload(), StandardCharsets.UTF_8),
+                                                          NotificationMessage.class);
+              try {
+                LOG.trace("Decoded notification: {}", message);
+                notificationReceived(message.getFeedId(), message.getNotificationJson());
+              } catch (Throwable t) {
+                LOG.warn("Error while processing notification {} with handler {}", message, t);
+              }
+              messageId = rawMessage.getId();
+            }
+          }
+        } catch (Exception e) {
+          LOG.error("Failed to get notification", e);
+        }
+
+        // Back-off if it was empty fetch.
+        if (emptyFetchCount > 0) {
+          // Schedule the next fetch. Exponential strategy doesn't use the time component,
+          // so doesn't matter what we passed in
+          subscribeExecutor.schedule(this,
```

Comment:
Let's say the first time we try to fetch messages, nothing is available. So we schedule our next execution, but meanwhile, a few notifications are published. How will we fetch those notifications, if we use the System#currentTimeMillis again during the next execution?